cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll

```diff
@@ -59,7 +60,70 @@ class MatchValue extends AbstractValue, TMatchValue {
 }
 
 /**
- * A Boolean condition in the AST that guards one or more basic blocks.
+ * A value number such that at least one of the instructions is
+ * a `CompareInstruction`.
+ */
+private class CompareValueNumber extends ValueNumber {
+  CompareInstruction cmp;
+
+  CompareValueNumber() { cmp = this.getAnInstruction() }
+
+  /** Gets an `CompareInstruction` belonging to this value number. */
+  CompareInstruction getCompareInstruction() { result = cmp }
+
+  /**
+   * Gets the left and right operands of a `CompareInstruction` that
+   * belongs to this value number.
+   */
+  predicate hasOperands(Operand left, Operand right) {
+    left = cmp.getLeftOperand() and
+    right = cmp.getRightOperand()
+  }
+}
+
+private class CompareEQValueNumber extends CompareValueNumber {
+  override CompareEQInstruction cmp;
+}
+
+private class CompareNEValueNumber extends CompareValueNumber {
+  override CompareNEInstruction cmp;
+}
+
+private class CompareLTValueNumber extends CompareValueNumber {
+  override CompareLTInstruction cmp;
+}
+
+private class CompareGTValueNumber extends CompareValueNumber {
+  override CompareGTInstruction cmp;
+}
+
+private class CompareLEValueNumber extends CompareValueNumber {
+  override CompareLEInstruction cmp;
+}
+
+private class CompareGEValueNumber extends CompareValueNumber {
+  override CompareGEInstruction cmp;
+}
+
+/**
+ * A value number such that at least one of the instructions is an
+ * insruction that is used in a `SwitchInstruction`'s expression.
+ */
+private class ScrutineeValueNumber extends ValueNumber {
```

Comment:
Maybe `ControllingValueNumber`?

Reply:
I like that (mostly because I hate the word "scrutinee" 😭). The only thing I dislike about `Controlling` is that it doesn't make it clear that it's specifically for switches (and I think "scrutinee" is a slightly more standard way of describing the expression in a switch.

Reply:
So I was just going by how the comments in the IR refer to them. The standard just seems to call it a "condition" so how about `SwitchCondition`?

Reply:
<img src="https://charlestonengravers.com/wp-content/uploads/2020/04/PersonalizedWalnutGavel_Main.webp"/>

Reply:
Fixed in 5f9d3d7b09270c33eed40a0994344696d8c92ad2